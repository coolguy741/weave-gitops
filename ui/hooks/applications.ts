import { useContext, useState } from "react";
import { AppContext } from "../contexts/AppContext";
import { Application } from "../lib/api/applications/applications.pb";

const WeGONamespace = "wego-system";

export default function useApplications() {
  const { applicationsClient, doAsyncError } = useContext(AppContext);
  const [loading, setLoading] = useState(true);

  const listApplications = (namespace: string = WeGONamespace) => {
    setLoading(true);

    return applicationsClient
      .ListApplications({ namespace: namespace })
      .then((res) => res.applications)
      .catch((err) => doAsyncError(err.message, err.detail))
      .finally(() => setLoading(false));
  };

  const getApplication = (name: string) => {
    setLoading(true);

    return applicationsClient
      .GetApplication({ name, namespace: WeGONamespace })
      .then((res) => res.application)
      .catch((err) => doAsyncError("Error fetching application", err.message))
      .finally(() => setLoading(false));
  };

  const listCommits = (app: Application) => {
    return applicationsClient.ListCommits({ ...app });
  };

  return {
    loading,
    listApplications,
    listCommits,
    getApplication,
  };
}
