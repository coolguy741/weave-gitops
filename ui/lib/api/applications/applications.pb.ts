/* eslint-disable */
// @ts-nocheck
/*
* This file is a generated Typescript file for GRPC Gateway, DO NOT MODIFY
*/

import * as fm from "./fetch.pb"
export type Application = {
  name?: string
}

export type ListApplicationsRequest = {
}

export type ListApplicationsResponse = {
  applications?: Application[]
}

export class Applications {
  static ListApplications(req: ListApplicationsRequest, initReq?: fm.InitReq): Promise<ListApplicationsResponse> {
    return fm.fetchReq<ListApplicationsRequest, ListApplicationsResponse>(`/v1/applications?${fm.renderURLSearchParams(req, [])}`, {...initReq, method: "GET"})
  }
}