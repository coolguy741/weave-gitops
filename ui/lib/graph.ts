// A collection of helper functions to render a graph of kubernetes objects
// with in the context of their parent-child relationships.
import {
  Applications,
  GroupVersionKind,
  UnstructuredObject,
} from "./api/applications/applications.pb";

export type UnstructuredObjectWithParent = UnstructuredObject & {
  parentUid?: string;
};

// Kubernetes does not allow us to query children by parents.
// We keep a list of common parent-child relationships
// to look up children recursively.
export const PARENT_CHILD_LOOKUP = {
  Deployment: {
    group: "apps",
    version: "v1",
    kind: "Deployment",
    children: [
      {
        group: "apps",
        version: "v1",
        kind: "ReplicaSet",
        children: [{ version: "v1", kind: "Pod" }],
      },
    ],
  },
};

export const getChildrenRecursive = async (
  appsClient: typeof Applications,
  result: UnstructuredObjectWithParent[],
  object: UnstructuredObjectWithParent,
  lookup: any
) => {
  result.push(object);

  const k = lookup[object.groupVersionKind.kind];

  if (k && k.children) {
    for (let i = 0; i < k.children.length; i++) {
      const child: GroupVersionKind = k.children[i];

      const res = await appsClient.GetChildObjects({
        parentUid: object.uid,
        groupVersionKind: child,
      });

      for (let q = 0; q < res.objects.length; q++) {
        const c = res.objects[q];

        // Dive down one level and update the lookup accordingly.
        await getChildrenRecursive(
          appsClient,
          result,
          { ...c, parentUid: object.uid },
          {
            [child.kind]: child,
          }
        );
      }
    }
  }
};
