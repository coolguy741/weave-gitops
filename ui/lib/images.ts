/* eslint-disable */
// @ts-nocheck

import failedSrc from "data-url:../images/failed.svg";
import fancyCardBackgroundBlue from "data-url:../images/fancy-card-background-blue.svg";
import fancyCardBackgroundOrange from "data-url:../images/fancy-card-background-orange.svg";
import fancyCardBackground from "data-url:../images/fancy-card-background.svg";
import bg from "data-url:../images/General-backgroundok.png";
import logoSrc from "data-url:../images/logo.svg";
import signInWheel from "data-url:../images/SignInWheel.svg";
import successSrc from "data-url:../images/success.svg";
import suspendedSrc from "data-url:../images/suspended.svg";
import titleSrc from "data-url:../images/Title.svg";
import weaveLogo from "data-url:../images/WeaveLogo.svg";

export default {
  bg,
  failedSrc,
  fancyCardBackground,
  fancyCardBackgroundBlue,
  fancyCardBackgroundOrange,
  logoSrc,
  signInWheel,
  successSrc,
  suspendedSrc,
  titleSrc,
  weaveLogo,
};
