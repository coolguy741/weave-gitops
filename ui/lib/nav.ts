import qs from "query-string";
import { SourceRefSourceKind } from "./api/core/types.pb";
import { NoNamespace, PageRoute, V2Routes } from "./types";

// getParentNavValue returns the parent for a child page.
// This keeps the nav element highlighted if we are on a child page.
// Example: /sources and /git_repo will both show the "Sources" nav as selected.
export const getParentNavValue = (
  currentPage: any
): PageRoute | V2Routes | boolean => {
  switch (currentPage) {
    case V2Routes.Automations:
    case V2Routes.Kustomization:
    case V2Routes.HelmRelease:
      return V2Routes.Automations;

    case V2Routes.Sources:
    case V2Routes.GitRepo:
    case V2Routes.HelmChart:
    case V2Routes.HelmRepo:
    case V2Routes.Bucket:
      return V2Routes.Sources;

    case V2Routes.FluxRuntime:
      return V2Routes.FluxRuntime;

    default:
      // The "Tabs" component of material-ui wants a bool
      return false;
  }
};

const pageTitles = {
  [V2Routes.Automations]: "Applications",
  [V2Routes.Sources]: "Sources",
  [V2Routes.FluxRuntime]: "Flux Runtime",
};

export const getPageLabel = (route: V2Routes): string => {
  return pageTitles[route];
};

export const formatURL = (page: string, query: any = {}) => {
  return `${page}?${qs.stringify(query)}`;
};

export const formatSourceURL = (
  kind: SourceRefSourceKind,
  name: string,
  namespace: string = NoNamespace
) => {
  return formatURL(sourceTypeToRoute(kind), { name, namespace });
};

export function sourceTypeToRoute(t: SourceRefSourceKind): V2Routes {
  switch (t) {
    case SourceRefSourceKind.GitRepository:
      return V2Routes.GitRepo;

    case SourceRefSourceKind.Bucket:
      return V2Routes.Bucket;

    case SourceRefSourceKind.HelmRepository:
      return V2Routes.HelmRepo;

    case SourceRefSourceKind.HelmChart:
      return V2Routes.HelmChart;

    default:
      break;
  }
}
