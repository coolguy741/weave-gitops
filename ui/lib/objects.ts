import { stringify } from "yaml";
import { FluxObjectKind, Object as ResponseObject } from "./api/core/types.pb";

export enum Kind {
  GitRepository = "GitRepository",
  Bucket = "Bucket",
  HelmRepository = "HelmRepository",
  HelmChart = "HelmChart",
  Kustomization = "Kustomization",
  HelmRelease = "HelmRelease",
}

export function fluxObjectKindToKind(fok: FluxObjectKind): Kind {
  return Kind[FluxObjectKind[fok].slice(4)];
}

export class FluxObject {
  obj: any;

  constructor(response: ResponseObject) {
    this.obj = JSON.parse(response.payload);
  }

  yaml() {
    return stringify(this.obj);
  }

  // Return list of key-value pairs for the metadata annotations that follow
  // our spec
  metadata(): [string, string][] {
    const prefix = "metadata.weave.works/";
    const annotations = this.obj.metadata.annotations || {};
    return Object.keys(annotations).flatMap((key) => {
      if (!key.startsWith(prefix)) {
        return [];
      } else {
        return [[key.slice(prefix.length), annotations[key] as string]];
      }
    });
  }
}
