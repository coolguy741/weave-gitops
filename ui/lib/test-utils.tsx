import { MuiThemeProvider } from "@material-ui/core";
import { createMemoryHistory } from "history";
import _ from "lodash";
import * as React from "react";
import { QueryClient, QueryClientProvider } from "react-query";
import { Router } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import AppContextProvider, { AppProps } from "../contexts/AppContext";
import { CoreClientContext } from "../contexts/CoreClientContext";
import {
  Applications,
  GetGithubAuthStatusRequest,
  GetGithubAuthStatusResponse,
  GetGithubDeviceCodeRequest,
  GetGithubDeviceCodeResponse,
  ParseRepoURLRequest,
  ParseRepoURLResponse,
  ValidateProviderTokenRequest,
  ValidateProviderTokenResponse,
} from "./api/applications/applications.pb";
import {
  Core,
  GetChildObjectsRequest,
  GetChildObjectsResponse,
  GetReconciledObjectsRequest,
  GetReconciledObjectsResponse,
  GetVersionRequest,
  GetVersionResponse,
  ListObjectsRequest,
  ListObjectsResponse,
} from "./api/core/core.pb";
import theme, { muiTheme } from "./theme";
import { RequestError } from "./types";

export type CoreOverrides = {
  GetChildObjects?: (req: GetChildObjectsRequest) => GetChildObjectsResponse;
  GetReconciledObjects?: (
    req: GetReconciledObjectsRequest
  ) => GetReconciledObjectsResponse;
  GetVersion?: (req: GetVersionRequest) => GetVersionResponse;
  ListObjects?: (req: ListObjectsRequest) => ListObjectsResponse;
};

export const createCoreMockClient = (
  ovr: CoreOverrides,
  error?: RequestError
): typeof Core => {
  const promisified = _.reduce(
    ovr,
    (result, handlerFn, method) => {
      result[method] = (req) => {
        if (error) {
          return new Promise((_, reject) => reject(error));
        }
        return new Promise((accept) => accept(handlerFn(req) as any));
      };

      return result;
    },
    {}
  );

  return promisified as typeof Core;
};

export type ApplicationOverrides = {
  GetGithubDeviceCode?: (
    req: GetGithubDeviceCodeRequest
  ) => GetGithubDeviceCodeResponse;
  GetGithubAuthStatus?: (
    req: GetGithubAuthStatusRequest
  ) => GetGithubAuthStatusResponse;
  ParseRepoURL?: (req: ParseRepoURLRequest) => ParseRepoURLResponse;
  ValidateProviderToken?: (
    req: ValidateProviderTokenRequest
  ) => ValidateProviderTokenResponse;
};

// Don't make the user wire up all the promise stuff to be interface-compliant
export const createMockClient = (
  ovr: ApplicationOverrides,
  error?: RequestError
): typeof Applications => {
  const promisified = _.reduce(
    ovr,
    (result, handlerFn, method) => {
      result[method] = (req) => {
        if (error) {
          return new Promise((_, reject) => reject(error));
        }
        return new Promise((accept) => accept(handlerFn(req) as any));
      };

      return result;
    },
    {}
  );

  return promisified as typeof Applications;
};

export function withTheme(element) {
  return (
    <MuiThemeProvider theme={muiTheme}>
      <ThemeProvider theme={theme}>{element}</ThemeProvider>
    </MuiThemeProvider>
  );
}

type TestContextProps = AppProps & { api?: typeof Core };

export function withContext(
  TestComponent,
  url: string,
  { api, ...appProps }: TestContextProps
) {
  const history = createMemoryHistory();
  history.push(url);
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  });
  const isElement = React.isValidElement(TestComponent);
  return (
    <Router history={history}>
      <AppContextProvider renderFooter {...appProps}>
        <QueryClientProvider client={queryClient}>
          <CoreClientContext.Provider value={{ api, featureFlags: {} }}>
            {isElement ? TestComponent : <TestComponent />}
          </CoreClientContext.Provider>
        </QueryClientProvider>
      </AppContextProvider>
    </Router>
  );
}
