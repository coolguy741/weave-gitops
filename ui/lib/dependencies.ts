import { Automation } from "../hooks/automations";
import { FluxObjectNode } from "./objects";

// findNode searches for a node by the node's name and namespace.
export function findNode(
  nodes: FluxObjectNode[],
  name: string,
  namespace: string
): FluxObjectNode | null {
  const matches = nodes.filter(
    (node) => node.name === name && node.namespace === namespace
  );

  if (matches.length > 0) {
    return matches[0];
  } else {
    return null;
  }
}

// getNeighborNodes returns nodes which depend on the current node
// or are dependencies of the current node.
export function getNeighborNodes(
  nodes: FluxObjectNode[],
  currentNode: FluxObjectNode
): FluxObjectNode[] {
  const dependencyNodes = currentNode.dependsOn
    .map((dependency) => {
      const name = dependency.name;
      let namespace = dependency.namespace;

      if (!namespace) {
        namespace = currentNode.namespace;
      }

      return findNode(nodes, name, namespace);
    })
    .filter((n) => n);

  const dependentNodes = nodes.filter((node) => {
    let isDependent = false;

    for (const dependency of node.dependsOn) {
      const name = dependency.name;
      let namespace = dependency.namespace;
      if (!namespace) {
        namespace = node.namespace;
      }

      if (name === currentNode.name && namespace === currentNode.namespace) {
        isDependent = true;
        break;
      }
    }

    return isDependent;
  });

  return dependencyNodes.concat(dependentNodes);
}

// getGraphNodes returns all nodes in the current node's dependency tree, including the current node.
export function getGraphNodes(
  nodes: FluxObjectNode[],
  automation: Automation
): FluxObjectNode[] {
  // Find node, corresponding to the automation.
  const currentNode = findNode(nodes, automation.name, automation.namespace);

  if (!currentNode) {
    return [];
  }

  currentNode.isCurrentNode = true;

  // Find nodes in the current node's dependency tree.
  let graphNodes: FluxObjectNode[] = [];

  const visitedNodes = new Map<string, boolean>();

  let nodesToExplore: FluxObjectNode[] = [currentNode].concat(
    getNeighborNodes(nodes, currentNode)
  );

  if (nodesToExplore.length === 1) {
    return [];
  }

  while (nodesToExplore.length > 0) {
    const node = nodesToExplore.shift();

    const newNodes = getNeighborNodes(nodes, node).filter(
      (n) => !visitedNodes[n.id]
    );

    for (const n of newNodes) {
      visitedNodes[n.id] = true;
    }

    nodesToExplore = nodesToExplore.concat(newNodes);

    graphNodes = graphNodes.concat(node);
  }

  // Remove duplicates from graphNodes.
  graphNodes = graphNodes.filter(
    (node, index) => graphNodes.indexOf(node) === index
  );

  return graphNodes;
}
