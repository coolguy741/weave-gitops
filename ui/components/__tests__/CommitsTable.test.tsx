import { render, screen } from "@testing-library/react";
import "jest-styled-components";
import * as React from "react";
import { unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { ListCommitsResponse } from "../../lib/api/applications/applications.pb";
import { withContext, withTheme } from "../../lib/test-utils";
import CommitsTable from "../CommitsTable";

describe("CommitsTable", () => {
  let container = null;
  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });
  describe("snapshots", () => {
    it("renders", async () => {
      const ovr = {
        ListCommits: (): ListCommitsResponse => ({
          commits: [
            {
              hash: "123abc",
              author: "Example User",
              date: "2021-09-10T23:45:09Z",
            },
          ],
        }),
      };
      const app = { name: "my-app", namespace: "my-ns" };
      let c;
      await act(async () => {
        const { container: div } = render(
          withTheme(
            withContext(<CommitsTable app={app} authSuccess={true} />, "/", ovr)
          ),
          container
        );
        c = div;
      });
      expect(c.firstChild).toMatchSnapshot();
    });
  });
  it("shows commits", async () => {
    const hash = "123abc";
    const commits = [
      { hash, author: "Example User", date: "2021-09-10T23:45:09Z" },
    ];
    const ovr = {
      ListCommits: (): ListCommitsResponse => ({
        commits,
      }),
    };
    const app = { name: "my-app", namespace: "my-ns" };

    await act(async () => {
      render(
        withTheme(
          withContext(<CommitsTable app={app} authSuccess={true} />, "/", ovr)
        ),
        container
      );
    });
    const row = await screen.findByText(hash);
    expect(row.innerHTML).toEqual(hash);
  });
});
