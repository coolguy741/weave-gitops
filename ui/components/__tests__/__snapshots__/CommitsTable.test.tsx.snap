// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommitsTable snapshots renders 1`] = `
.c1 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 16px;
  font-weight: 400;
  text-transform: none;
  font-style: normal;
}

.c3 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 16px;
  font-weight: 400;
  text-transform: none;
  font-style: normal;
  color: #00b3ec;
}

.c2 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2.title-bar-button {
  width: 250px;
}

.c0 .MuiTableCell-head {
  font-weight: 800;
}

<div
  class=""
>
  <div
    class="c0"
  >
    <div
      class="MuiTableContainer-root"
    >
      <table
        aria-label="simple table"
        class="MuiTable-root"
      >
        <thead
          class="MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head"
              scope="col"
            >
              SHA
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head"
              scope="col"
            >
              Date
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head"
              scope="col"
            >
              Message
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head"
              scope="col"
            >
              Author
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              <span
                class="c1"
              >
                <a
                  class="c2"
                  href="/"
                >
                  <span
                    class="c3"
                    color="primary"
                  >
                    123abc
                  </span>
                </a>
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              <span
                class="c1"
              >
                test timestamp
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              <span
                class="c1"
              />
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              <span
                class="c1"
              >
                Example User
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
